src/lib.mo

```diff
@@ -1,10 +1,168 @@
-/// ...
+/// This module implements an array of linked lists, fully stored in memory regions.
+/// All data is managed using two growable regions: `indexTable` and `data`.
+
+/// To use a list, client code must first allocate it using `allocateList`.
+/// Each list is identified by a unique index — a consecutive natural number starting from 0.
+
+/// --- INDEX TABLE REGION ---
+/// The index table stores metadata for each list. Each record is 24 bytes long,
+/// and the record for list `N` starts at offset `24 * N`.
+
+/// Index table record structure:
+/// | Offset | Type  | Description                                                          |
+/// |--------|-------|----------------------------------------------------------------------|
+/// | 0      | Nat64 | Number of items in the list                                          |
+/// | 8      | Nat64 | Pointer (offset in `data`) to the first item in the list (0 if none) |
+/// | 16     | Nat64 | Pointer (offset in `data`) to the last item in the list (0 if none)  |
+
+/// --- DATA REGION ---
+/// The data region stores the actual list items.
+/// It begins with a single reserved byte so that pointer value `0` can be used as a null pointer.
+
+/// Each item is stored as a record with the following structure:
+/// | Offset | Type  | Description                                          |
+/// |--------|-------|------------------------------------------------------|
+/// | 0      | Nat64 | Pointer to the previous item in the list (0 if none) |
+/// | 8      | Nat64 | Pointer to the next item in the list (0 if none)     |
+/// | 16     | Nat16 | Size of the data blob                                |
+/// | 18     | Blob  | Actual item data                                     |
 ///
-/// Copyright: 2023-2024 MR Research AG
-/// Main author: 
-/// Contributors: 
+/// Copyright: 2023-2025 MR Research AG
+/// Main author: Andy Gura
+/// Contributors: Timo Hanke
 
+import Blob "mo:base/Blob";
+import Iter "mo:base/Iter";
+import Nat16 "mo:base/Nat16";
+import Nat64 "mo:base/Nat64";
 import Prim "mo:prim";
+import Region "mo:base/Region";
 
 module {
+
+  public type StableLogLists = {
+    indexTable : Region;
+    data : Region;
+    var dataLength : Nat;
+    var listsAmount : Nat;
+    var totalRecords : Nat;
+  };
+
+  type Record = { prevPtr : Nat64; nextPtr : Nat64; data : Blob };
+
+  public func new() : StableLogLists = {
+    indexTable = Region.new();
+    data = Region.new();
+    var dataLength = 1;
+    var listsAmount = 0;
+    var totalRecords = 0;
+  };
+
+  public func size(l : StableLogLists, listIndex : Nat) : Nat = loadListLength_(l, listIndex) |> Nat64.toNat(_);
+
+  public func totalSize(l : StableLogLists) : Nat = l.totalRecords;
+
+  public func allocateList(l : StableLogLists) : Nat {
+    let newlistIndex = l.listsAmount;
+    if (65536 * Region.size(l.indexTable) < (Nat64.fromNat(newlistIndex) + 1) * 24 and Region.grow(l.indexTable, 1) == 0xFFFF_FFFF_FFFF_FFFF) {
```

Comment:
I think nesting this with two `if`'s is clearer. I mean separate the condition for growing and the error condition into two lines.